Prune pipelines / pipeline runs

michaelsauter opened the pull request:
Closes #153.

Tasks: 
- [x] Updated design documents in `docs/design` directory or not applicable
- [x] Updated user-facing documentation in `docs` directory or not applicable
- [x] Ran tests (e.g. `make test`) or not applicable
- [x] Updated changelog or not applicable

gerardcl reviewed (approved):
so far it LGTM! If I got it right, old pipelines already existing prior upgrading to a new ods-pipeline release (that includes this PR) will need manual cleanup, right?

I added some small comments and questions and maybe I would go for adding another test with changing to different edge/expected cases with the params `minKeepHours` and `maxKeepRuns` in `pruner_test.go`?

but approving for now! this is looking good!! 🚀

michaelsauter commented on cmd/start/main.go:
```diff
@@ -200,7 +200,7 @@ func main() {
 	}
 
-	if len(ctxt.Environment) > 0 {
+	if ctxt.Environment != "" {
```
No, both will always work. Go never has "undefined" values, but every type has a "zero value". `ctxt.Environment` is a `string`, and the zero value of a string is `""`. It is idiomatic Go to compare with the zero value if you want to check if sth. is set. Hence, I changed to string comparison instead of length, even though both works exactly the same!

michaelsauter commented on internal/interceptor/pruner.go:
```diff
@@ -0,0 +1,190 @@
+// Prune prunes runs within pipelineRuns which can be cleaned up according to
+// the strategy in pipelineRunPrunerByStage.
+func (p *pipelineRunPrunerByStage) Prune(ctxt context.Context, pipelineRuns []tekton.PipelineRun) error {
+	prByStage := p.categorizePipelineRunsByStage(pipelineRuns)
```
hehe, I think I'll leave it for now. Tekton uses `pr` often to abbreviate pipeline run so I think we can do it too :)

michaelsauter commented on internal/interceptor/pruner.go:
```diff
@@ -0,0 +1,190 @@
+}
+
+// unique returns a slice of strings where all items appear only once.
+func unique(stringSlice []string) []string {
```
Totally. We do not have it elsewhere and do not need it elsewhere right now. I'll leave it here and sometime in the future(Go 1.19?) we may be able to replace it with a stdlib func...

michaelsauter commented:
> If I got it right, old pipelines already existing prior upgrading to a new ods-pipeline release (that includes this PR) will need manual cleanup, right?

Correct, I added a comment in the changelog regarding this.

> I added some small comments and questions and maybe I would go for adding another test with changing to different edge/expected cases with the params minKeepHours and maxKeepRuns in pruner_test.go?

Good point! I added some validation to ensure we have at least minKeepHours>=1 and maxKeepRuns>=1. The logic would work also for minKeepHours=0 and maxKeepRuns=0, but I think that is unlikely to be what the user actually wanted to do, so better to prevent it.
